app/services/solidus_pay_tomorrow/client/base_service.rb

```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+
+require 'httparty'
+
+module SolidusPayTomorrow
+  module Client
+    class BaseService
+      attr_reader :payment_method
+
+      STAGING_URL = 'https://api-staging.paytomorrow.com'
+      PRODUCTION_URL = 'https://api.paytomorrow.com'
+      AUTH_ENDPOINT = 'api/uaa/oauth/token'
+
+      def initialize(payment_method:, **_args)
+        @payment_method = payment_method
+      end
+
+      class << self
+        attr_accessor :current_token
+
+        def call(*args, **kwargs, &block)
+          new(*args, **kwargs, &block).call
+        end
+      end
+
+      private
+
+      def api_base_url
+        if payment_method.preferred_test_mode
+          STAGING_URL
+        else
+          PRODUCTION_URL
+        end
+      end
+
+      def auth_headers
+        { 'Authorization': "Bearer #{valid_token}",
+          'Content-Type': 'application/json' }
+      end
+
+      def handle_errors!(result)
+        return result.parsed_response if result.success?
+
+        raise StandardError, result.parsed_response['error']
+      end
+
+      def valid_token
+        new_token if current_token_invalid?
+        self.class.current_token[:access_token]
+      end
+
+      def new_token
+        result = handle_errors!(HTTParty.post(token_uri, headers: token_headers, body: URI.encode_www_form(token_body)))
+        self.class.current_token =
+          { expires_at: Time.zone.now + result['expires_in'].seconds }.merge!(result.symbolize_keys)
+      end
+
+      # Tokens are valid for 10-12 hours, if they are near expiry, get new token
+      def current_token_invalid?
+        self.class.current_token.nil? || Time.zone.now > self.class.current_token[:expires_at] - 10.minutes
```

Comment:
Why this refreshes the token 10 minutes before the expiration? 
I think we just need to refresh it when it is expired.

Reply:
Hmmm, I am just thinking of a race condition where this(_Time.zone.now > self.class.current_token[:expires_at]_) condition returns false during code execution, it becomes true just after and we do the request with expired token.
**Example** - current token is going to be expired 1 second after executing this code.

Hence, I think getting a new token with a margin would be better, wdyt?

Although, this time gap would be in < a second maybe so I think instead of 10 minutes, 1 minute should be fine?

Reply:
It makes sense. We can leave 10 minutes for it. Thank you for the explaination